File: examples/websocket_loopback.rs
//! A WebSocket echo server and a client, in one process over loopback TCP.
//!
//! Needs no network access and no certificate:
//!
//! ```bash
//! cargo run --example websocket_loopback
//! ```

use soyokaze::models::Port;
use soyokaze::protocol::base::Transport;
use soyokaze::websocket::{CloseCode, Opcode, WebSocketConnection};
use soyokaze::{Client, Handler, Server};

struct Echo;

impl Handler for Echo {
    async fn on_websocket(&self, mut socket: WebSocketConnection<Box<dyn Transport>>) {
        while let Ok((opcode, payload)) = socket.receive_message().await {
            if socket.send_message(opcode, payload.to_vec()).await.is_err() {
                break;
            }
        }

        socket.close(CloseCode::Normal, "").await;
    }
}

#[tokio::main]
async fn main() -> Result<(), soyokaze::Error> {
    let server = Server::default();
    let handle = server.serve(Echo, &[Port::TCP(0)]).await?;
    let address = handle.address().expect("the server bound no address");

    let client = Client::default();
    let mut socket = client.websocket(&format!("ws://{address}/echo")).await?;

    for message in ["hello", "soyokaze"] {
        socket.send_message(Opcode::Text, message.as_bytes().to_vec()).await?;
        let (opcode, payload) = socket.receive_message().await?;

        println!("{opcode:?} {}", String::from_utf8_lossy(&payload));
    }

    socket.close(CloseCode::Normal, "").await;
    handle.close(Some(5.0)).await;
    Ok(())
}

File: python/examples/websocket_loopback.py
"""A WebSocket echo server and a client, in one process over loopback TCP.

The Python half of the crate's ``examples/websocket_loopback.rs``. It needs no
network access and no certificate:

    uv run examples/websocket_loopback.py
"""

import asyncio

from soyokaze import Client, CloseCode, Error, Message, Opcode, Port, Server

async def echo(socket):
    """Sends every message back the way it came, until the socket ends.

    The handler drives the socket for as long as it lives, awaiting the next
    message rather than holding a thread while nothing arrives; the bindings
    close and free it once the handler returns, so this one simply stops when
    the connection does.
    """
    try:
        while True:
            opcode, payload = await socket.receive_message()
            await socket.send_message(opcode, payload)
    except Error:
        pass

async def decline(request):
    """Answers every request that was not a WebSocket upgrade."""
    response = Message.response(426, request.version)
    response.set_body("this port speaks WebSocket")
    return response

async def main():
    # A port of zero lets the kernel choose one, so the example names none.
    server = Server()
    handle = await server.serve(decline, [Port.TCP(0)], on_websocket=echo)

    client = Client()
    socket = await client.websocket(f"ws://127.0.0.1:{handle.port}/echo")

    for message in ["hello", "soyokaze"]:
        await socket.send_message(Opcode.TEXT, message)
        opcode, payload = await socket.receive_message()
        print(f"{opcode.name} {payload.decode()}")

    await socket.close(CloseCode.NORMAL, "")
    await handle.close(5)

if __name__ == "__main__":
    asyncio.run(main())
